Add guidelines for writing inline docstrings to the manual

nalimilan opened the pull request:
Now that we have some experience with writing docstrings, it looks like a good time to try providing some guidance about how to organize them (a question which emerged at https://github.com/JuliaLang/DataStructures.jl/pull/182). So here's a proposal for discussion.

nalimilan committed:
Synchronize documentation indexes

Also add comments to prevent them from getting out of sync again.

nalimilan commented:
Related: https://github.com/JuliaLang/julia/issues/8966

This PR is more restricted in its scope, as it merely attempts to standardize existing practices. But the part about the argument lists section could certainly be extended to list other sections in the future.

nalimilan commented on doc/manual/documentation.rst:
```diff
@@ -18,25 +24,66 @@ In 0.4:
+  purpose. An argument list would only repeat information already provided elsewhere.
+  However, it it might be a good idea for complex functions with many arguments. In
+  that case, put the argument list after the general description of the function, under
+  a ``# Arguments:`` header, with one ``*`` bullet for each argument.
```
Currently, it makes more sense to use `### Arguments:`, since higher-level headings are too heavy. But it doesn't make much sense IMHO that we would need to use third-level headings for sections, which are arguably top-level for a docstring. So maybe top-level headings should be presented in a less heavy way (i.e. with a single underline)?

MichaelHatherly replied:
Agreed, rust uses that style (https://doc.rust-lang.org/book/documentation.html#special-sections), and it reads quite nicely I think. Maybe dispense with the `:` at the end of the title?

We'd need to do a little bit of re-styling of the terminal output, as you mention, and the style for generated docs, but that's not much of a roadblock.

nalimilan replied:
Hey, you're fast! So let's keep it as `#` and expect the display to change. The important point is to have some consistency in the definition of sections.

I agree the `:` can be left out, will fix.

hayd replied:
I had written something before (in Lexicon IIRC) which normalized the headers in a block e.g. all be at most h3. If there's an h1, all get 2 subtracted etc.

nalimilan replied:
Could be a good solution indeed. At least it would avoid breaking existing documentation.

MichaelHatherly commented on doc/manual/documentation.rst:
```diff
@@ -18,25 +24,66 @@ In 0.4:
-Markdown happens to be the default, but one can construct other string
-macros and pass them to the ``@doc`` macro just as well.
+1. Always show the signature of a function at the top of the documentation,
+with a four-space ident so that it is printed as Julia code.
```
"ident" -> "indent"

nalimilan replied:
Ah, I always make this mistake. Will fix.

nalimilan commented on doc/manual/documentation.rst:
```diff
@@ -18,25 +24,66 @@ In 0.4:
+    the index between all pairs of columns of `x`.
 
-        @bar buy_drink_for("Jiahao")
+    Example:
```
I realize this should be another section, will change to `# Examples`.

nalimilan commented:
Rust also has a RFC with detailed guidelines like the ones I wrote here: https://github.com/rust-lang/rfcs/blob/master/text/0505-api-comment-conventions.md
It also mentions punctuation and persons to use for verbs. I could add this to the PR if people think it's a good idea (IMO, anything that makes documentation writers wonder less about how to best write it is a good idea...).

mauro3 commented on doc/manual/documentation.rst:
```diff
@@ -18,25 +24,67 @@ In 0.4:
+  that case, put the argument list after the general description of the function, under
+  a ``# Arguments`` header, with one ``*`` bullet for each argument.
+
+4. When adding examples, use ``jldoctest`` blocks instead of standard text.
```
What is `jldoctest`?  How do I run doctests?  Needs some explanation, maybe combined with something in the Unit Tests section.

nalimilan replied:
Actually, I need somebody to tell me whether/how to run doctests. Even if it's not yet possible, it would still be a good idea to advise people to use them so that everything is ready when support lands.

hayd replied:
At the moment this is base only cc @MichaelHatherly

MichaelHatherly replied:
Yes, base only. I believe they are run with sphinx. I'd like to eventually just have any `julia` code block get tested without having to manually add a doctest flag to code blocks, though that might be too aggressive perhaps. Checking whether the block contains any REPL prompts, i.e. `julia> `, and running those may be a suitable heuristic. There's some work done in this direction in Lapidary.jl.

nalimilan replied:
So should I advise using `julia` blocks with `julia>` prompts for now (and only indicate that in the future it will be used for automated testing)?

mauro3 commented:
Thanks, look like reasonable suggestions to me.

DanielArndt commented:
@nalimilan Is it worth putting something about line length in (ie. breaking up large paragraphs into lines of 80 chars)? Seems that it is a common habit to put in one long line.

nalimilan commented:
Indeed, makes sense.

omus commented:
What is the recommended line length with docstrings? [CONTRIBUTING.md](https://github.com/JuliaLang/julia/blob/master/CONTRIBUTING.md) recommends a 92 character limit for code.

nalimilan commented:
I'd say we should follow the same rule, since docstrings are mixed with code and edited using the same tools.

DanielArndt commented:
makes sense! I missed the 92 char limit. Anyone know significance of 92?

hayd commented on doc/manual/documentation.rst:
```diff
@@ -18,25 +24,67 @@ In 0.4:
+  This allows running examples automatically and checking that their actual output is
+  consistent with that presented in the documentation. This way, the code is tested and
+  examples cannot get out of date without notice. Examples should be grouped under an
+  ``# Examples`` section.
```
I had thought base / packages were using `###` or `####`, IMO h1 is very high.

nalimilan replied:
See now hidden discussion at https://github.com/JuliaLang/julia/pull/15136#discussion_r53326703

H1 is high, but within a docstring I don't see what could be higher than a section title...

hayd replied:
Tganks!

it's true, IMO the issue comes if you use all of h1-6, but that's an edge case that can be warned for.

nalimilan replied:
6 levels for a docstrings would be a lot. My idea was rather to avoid the noise of using three # when one would be enough, as well as the oddity of starting headings at level 3.

MichaelHatherly replied:
> H1 is high, but within a docstring I don't see what could be higher than a section title...

Yes, the docstring should be able to act as a standalone "document", so each major section of the "document" is in fact a toplevel section regardless of whether it might be embedded in some other larger document at some point.

The noise argument is also a compelling one for a single `#`.

Also possibly worth mentioning whether to use the plural `# Examples` even if there's only one example. I'd be in favour of sticking to always using plural form.

jiahao replied:
There's a serious formatting incompatibility with Sphinx in allowing section headings in docstrings, at least for `Base`. We currently put extracted docstrings back into the stdlib docs where they go under function blocks, but Sphinx complains with:

```
SEVERE: Unexpected section title.
```

the moment you try to use any section headings whatsoever in a docstring that gets spliced into the docs.

What's the fix for this problem?

nalimilan replied:
I guess they could be turned into standard bold font as a workaround.

MichaelHatherly replied:
That's a bit unfortunate. Post-processing in `genstdlib.jl` to convert headers to bold might be a workable solution for the moment.

MichaelHatherly commented:
> 92 char limit

When I did the last rst->md conversion that's the style I stuck to. Probably best to just have a single line limit for both docs and code rather than having two different ones.

Some other thoughts that could be relevant:

**Formatting:**

Give docstrings ample spacing to aid readability since vertical space isn't at a premium. Place the starting and ending `"""` characters on lines by themselves:

```julia
"""
...

...
"""
func(x, y) = ...
```

rather than

```julia
"""...

..."""
func(x, y) = ...
```

to make it more clear where docstrings start and end.

**Documenting functions/methods:**

When a module extends a generic function from another module avoid adding new docs for every
single new method. The behaviour of new methods should follow the behaviour of other methods
of the function and so the function's docstring should be sufficient for describing it's
behaviour. If a method diverges significantly from the expected behaviour then it should
probably be called something else.

Prefer documenting which functions a type extends in the type's docstring. So

```julia
"""
...

# Extends

- `Other.func`
- ...

...
"""
type T end

Other.func(::T) = ...
```

rather than

```julia
"""
...

...
"""
type T end

"""
Calls `func` on a `T`...
"""
Other.func(::T) = ...
```

(Not sure how well this approach would work in practice though, or whether it would be any better.)

**Summary Line:**

After the simplified signature block include a single one line sentence describing what the
object does or what it represents. If it cannot be easily summarised then splitting it into
separate compose-able parts *could* be beneficial. This should not be taken as an absolute
requirement for every single case though.

See for example the docstring for `push!`:

```julia
"""
    push!(collection, items...) -> collection

Insert one or more `items` at the end of `collection`.

...
"""
...
```

**LaTeX syntax:**

Perhaps worth mentioning double backtick syntax and `math` code blocks for writing equations. Using unicode where possible in these equations, so

```julia
"""
... ``α = 1`` ...
"""
```

rather than

```julia
"""
... ``\\alpha  = 1`` ...
"""
```

tkelman commented:
Is auto-populating signatures, at least for simple cases, still on the radar somehow?

jiahao commented on doc/manual/documentation.rst:
```diff
@@ -18,25 +24,67 @@ In 0.4:
+    Computes the Bar index between `x` and `y`. If `y` is missing, the function computes
+    the index between all pairs of columns of `x`.
 
-        @bar buy_drink_for("Jiahao")
```
:broken_heart:

nalimilan replied:
Yeah, sorry, I wanted a more complex example with optional arguments and a detailed description of the parameters. Maybe add an optional vector of beer types?

More seriously, I realize it would make sense to add the types to the signature, since that's what it advised below.

nalimilan commented:
@MichaelHatherly Good points, I'll add these.

Though about extending methods, the PR already mentions that one shouldn't repeat the function's docstring. I'm not sure about listing extended methods in the type's documentation: that's quite verbose, and `methodswith` give you this information if you want it.

@tkelman I think that's https://github.com/JuliaLang/julia/issues/13504 and https://github.com/JuliaLang/julia/pull/9838.

MichaelHatherly commented:
> Though about extending methods...

Yup, sure. I'm not terribly attached to the idea. What you've added about not repeating docs should be good enough for now. Thanks.

> Is auto-populating signatures, at least for simple cases, still on the radar somehow?

Yes, it is. Hopefully before 0.5 feature freeze... whenever that is.

biogeo commented:
Regarding headings and extending methods, I notice that at the moment, Julia's help system will display the docstrings for all methods of a function. So if you enter

    julia> ?*

you get the docstrings for multiplication, string concatenation, and matrix multiplication, one after the other, and the call signatures are the only thing delimiting them. If there is other Julia code in the docstring, then the call signatures don't stand out at all. The H1 level seems like a natural choice to delimit different methods' docstrings.  How about recommending use of the call signature with the H1 level as the first line of a docstring?

Quick example at the Julia prompt:

    julia> "# `foo(bar)`\n Foo a bar" foo(bar) = bar;
    julia> "# `foo(bar,baz)`\nFoo a bar with a baz" foo(bar,baz) = bar+baz;
    help?> foo
    search: foo floor ifloor pointer_from_objref OverflowError RoundFromZero
    
         foo(bar,baz)
        ≡≡≡≡≡≡≡≡≡≡≡≡≡≡
    
      Foo a bar with a baz
    
         foo(bar)
        ≡≡≡≡≡≡≡≡≡≡
    
      Foo a bar

(Pretty cyan text for code highlighting the call signatures not shown here.)

The other nice feature is this still looks good if you get help for a specific method:

    help?> foo(Any)
         foo(bar)
        ≡≡≡≡≡≡≡≡≡≡
    
      Foo a bar

I think it also looks quite good in an IJulia notebook.

I agree that in many or most cases a separate docstring for each signature is unwarranted, but for those cases in which it is good to have, this seems like a good solution to me.

nalimilan commented:
@biogeo I think this would be better handled directly by the printing method. For example, it could add some line breaks, or a horizontal separator, to make different methods visually distinct. That way, we don't add visual noise to functions with only one docstring.

nalimilan commented:
I've just pushed an updated commit which should take into account all remarks. Please tell me what you think.

bicycle1885 commented:
I really love the stylistic consistency of one-line docstrings in Python, which is specified in "PEP 0257 -- Docstring Conventions" <https://www.python.org/dev/peps/pep-0257/#one-line-docstrings>. Especially, I think it would be better to import the idea of the following note:

> The docstring is a phrase ending in a period. It prescribes the function or method's effect as a command ("Do this", "Return that"), not as a description; e.g. don't write "Returns the pathname ...".

omus commented on doc/manual/documentation.rst:
```diff
@@ -18,25 +24,113 @@ In 0.4:
+  Optional arguments should be put into brackets ``[]``, with their default values
+  when applicable. An alternative solution is to use several lines: one without
+  optional arguments, the other(s) with them. This solution can also be used to
+  document several related methods of a given function.
```
I think we need to be more specific and give examples as to how optional arguments should be displayed. Currently Base is very inconsistent with how this should be used: `open(filename, [mode])` vs. `accept(server[,client])`

Some examples taken directly from base:
```
DateTime(y[, m, d, h, mi, s, ms])
floor([T,] x, [digits, [base]])
digits([T], n, [base], [pad])
sort(v, [alg=<algorithm>,] [by=<transform>,] [lt=<comparison>,] [rev=false])
reverse(v [, start=1 [, stop=length(v) ]] )
```

I would propose that the above examples would be standardized to: 
```
DateTime(y, [m, d, h, mi, s, ms])
floor([T], x, [digits, base])
digits([T], n, [base, pad])
sort(v; [alg=<algorithm>, by=<transform>, lt=<comparison>, rev=false])
reverse(v, [start=1, stop=length(v)])
```

omus replied:
After some more thought and some discussion I think the following would be a better standard even though it is a bit more verbose:
```julia
DateTime(y, [m, [d, [h, [mi, [s, [ms]]]]]])
floor([T], x, [digits, [base]])
digits([T], n, [base, [pad]])
sort(v; [alg=<algorithm>], [by=<transform>], [lt=<comparison>], [rev=false])
reverse(v, [start=1, [stop=length(v)]])
```

tkelman replied:
I personally think using real julia syntax for this would be clearer than square brackets. There's precedent for brackets with optional args, but it's not real syntax.

omus replied:
@tkelman what about a pseudo-julia syntax?
```julia
DateTime(y, m=1, d=1, h=0, mi=0, s=0, ms=0)
floor(T=typeof(x), x, digits=0, base=10)
digits(T=typeof(n), n, base=10, pad=0)
sort(v; alg=<algorithm>, by=<transform>, lt=<comparison>, rev=false)
reverse(v, start=1, stop=length(v))
```

spurll replied:
I concur with @omus here. I'm fine with an alternate syntax like the one Python uses, though:

```julia
DateTime(y [, m [, d [, h [, mi [, s [, ms]]]]]])
floor([T,] x [, digits [, base]])
digits([T,] n [, base [, pad]])
sort(v; [alg=<algorithm>,] [by=<transform>,] [lt=<comparison>,] [rev=false])
reverse(v [, start=1 [, stop=length(v)]])```

biogeo replied:
I think the nested square brackets add way too much noise and make the documentation harder to read.  When an unambiguous form of each call signature is needed, it can always be retrieved with `methods(f)`.  In my opinion, the signature in the docstring should provide an at-a-glance overview of using the function, and a mnemonic for the user to be reminded of argument positions.  I think Matlab's general approach is a good guide here.

biogeo commented:
I think the nested square brackets add way too much noise and make the documentation harder to read.  When an unambiguous form of each call signature is needed, it can always be retrieved with `methods(f)`.  In my opinion, the signature in the docstring should provide an at-a-glance overview of using the function, and a mnemonic for the user to be reminded of argument positions.  I think Matlab's general approach is a good guide here.

nalimilan commented:
I've updated the PR with attempts to address the points raised above regarding optional arguments. Indeed, the best solution when a default value exists is to show it using the Julia syntax. When it does not exists, the text now recommends using nested brackets and gives examples. Hopefully deeply nested brackets will be rare, but I don't see a way of conveying that information in a more readable fashion, except maybe by using one line for each variant.

I've also added a word about listing keyword arguments when there are many of them. We had this issue recently regarding `readtable` in DataFrames.jl. Please also tell me what you think.

@bicycle1885 I've retained your suggestion, but the current documentation doesn't follow it at all. If you like it, a pull request to make the docs more consistent could be useful.

Is this good to merge now? Anyway, we can always improve this later.

omus commented on doc/manual/documentation.rst:
```diff
@@ -5,38 +5,140 @@
+    function bar(x, y) ...
+
+Documentation is very free-form, though we recommend following some simple
+conventions, illustrated in the above example.
```
Sentence could use a little rework. Maybe:
```
Documentation, as illustrated above, is very free-form though we recommend following some simple conventions:
```

nalimilan replied:
Good catch. I've even removed the mention of "free-form", since the whole point of this section is now to give guidelines.

omus commented:
Nice work @nalimilan.

bicycle1885 commented:
@nalimilan Thank you for considering it. I'll send a pull request if I find inconsistency.

nalimilan commented:
Any more objections/ideas, or is it OK to merge?

omus commented:
:+1:

nalimilan commented:
I'll merge soon if nobody objects. I guess this also needs backporting to the 0.4 manual.

tkelman commented on doc/manual/documentation.rst:
```diff
@@ -5,38 +5,139 @@
+  (like ``mean(x::AbstractArray)``), or a simplified form.
+  Optional arguments should be represented with their default values (i.e. ``f(x, y=1)``)
+  when possible, following the actual Julia syntax. Optional arguments which
+  do not have a default value should be put brackets (i.e. ``f(x[, y])`` and ``f(x[, y[, z]])``).
```
put in brackets

nalimilan replied:
Good catch. I've just fixed it (I also changed `<arguments>` to `<keyword arguments>`).

nalimilan committed:
Add guidelines for writing inline docstrings to the manual

[av skip]

nalimilan commented:
See https://github.com/JuliaLang/julia/pull/15627 for backport.

nalimilan commented:
Does anybody know how to get
```
``\\alpha = 1``
```
to render correctly in Sphinx docs? The double backslash breaks everything.

jiahao commented:
`\\\alpha`?

nalimilan commented:
Unfortunately not, this gives three backslashes, and the backquotes are still outside the block. (To see what it looks like currently, look for `alpha` in http://docs.julialang.org/en/latest/manual/documentation/.)

omus commented:
@nalimilan in the RST files you need to do 

    :math:`\alpha = 2`

omus commented:
I can't get it to render as an inline code block. It seems like the only way to get that to render correctly is to use a code block indented by four-spaces.

omus commented:
@nalimilan there is probably a better RST way but this work:

    :samp:`\`\`\\\\alpha = 1\`\``

renders inline as

    ``\\alpha = 1``

nalimilan commented:
Thanks! I've added this fix to https://github.com/JuliaLang/julia/pull/15627 and I'll push it to master directly after the tests pass.
